Response errors should result in a Left of error

AlistairB opened the pull request:
Even when the response `a` can still be successfully parsed.

Closes #577

nalchevanidze commented on morpheus-graphql-client/changelog.md:
```diff
@@ -10,6 +10,7 @@
 
 - more likely to rebuild when a file loaded by `defineByDocumentFile` or
   `defineByIntrospectionFile` is changed
+- response errors should always result in a fetch failure [#577](https://github.com/morpheusgraphql/morpheus-graphql/issues/577)
```
you can mention yourself: @AlistairB

nalchevanidze commented on morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs:
```diff
@@ -0,0 +1,76 @@
+client :: IO (Either String TestQuery)
+client = fetch resolver ()
+
+test :: TestTree
```
now i think that, if we test errors then in ideal case we test both cases regular errors, and partial response

nalchevanidze commented on morpheus-graphql-client/src/Data/Morpheus/Client/Fetch.hs:
```diff
@@ -60,7 +60,8 @@ class Fetch a where
-      processResponse JSONResponse {responseData = Just x} = Right x
+      processResponse JSONResponse {responseData = Just x, responseErrors = Nothing} = Right x
+      processResponse JSONResponse {responseData = Just x, responseErrors = Just []} = Right x
       processResponse invalidResponse = Left (show invalidResponse)
```
what do you think, maybe we can use JSON "decode"  instead of regular show?

nalchevanidze replied:
will provide easier understandable error messages.

AlistairB replied:
So you might do something like:

```haskell
      processResponse JSONResponse {responseData = Just x, responseErrors = Nothing} = Right x
      processResponse JSONResponse {responseData = Just x, responseErrors = Just []} = Right x
      processResponse JSONResponse {responseErrors = err} = Left $ decodeUtf8 $ encode err
```

but you can't do

```haskell
     processResponse invalidResponse = Left $ decodeUtf8 $ encode invalidResponse
```

As `JSONResponse` has an `a` and based on the `Fetch` typeclass we don't know that we have `ToJSON a` constraint. So you would need to either add the `ToJSON a` constraint, which maybe the user doesn't want to provide. Or exclude the `responseData` from the error message.

Based on this issue I think the current implementation is fine. Also, I think unless you prettify the json, a json error message is not too different to a `show` error message. What do you think?

AlistairB replied:
If you wanted to improve this more, I think the right solution would be to have `fetch` return a `m (Either FetchFailure a)` where (roughly sketched)

```haskell
data FetchFailure = 
  FetchConnectionIssue String
  | ResponseDecodeFailure String
  | ServerGQLErrors (NonEmpty GQLError)
  deriving (Show, ToJSON)
```

This gives the user structured data they can work with if they want, or they can just show / render to JSON. I'd suggest this be done as a separate issue/PR though.

AlistairB commented:
@nalchevanidze I've done 2/3 and left some comments on the other suggestion. Let know what you think. Thanks.

AlistairB commented:
@nalchevanidze I had a bit of time so I thought I'd implement something like https://github.com/morpheusgraphql/morpheus-graphql/issues/577#issuecomment-774554821

The missing piece is even though the `GQLError` is returned, it doesn't include arbitrary fields in the error response. ie.

```json
{
  "data": {
    "repository": null
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "repository"
      ],
      "locations": [
        {
          "line": 7,
          "column": 3
        }
      ],
      "message": "Could not resolve to a Repository with the name 'facebook/reacttt'."
    }
  ]
}
```

I'd like to be able to get the `type` field somehow.

I think to do this I'd like to update `GQLError` to be include arbitrary fields included in the response. However, this type is used through the various packages. I'd probably instead have something like:

```haskell
data FetchGQLError = FetchGQLError {
  fetchError :: GQLError,
  rawError :: HashMap Text Value
}
```

When you aren't busy with your thesis, I'm curious what you think of what I have implemented and this addition. Thanks

AlistairB commented:
@nalchevanidze I have updated the PR based on https://github.com/morpheusgraphql/morpheus-graphql/issues/577#issuecomment-823654783 . The only difference is I added another case to `FetchError` of `FetchErrorNoResult` where you get no `a` but also there are no errors. Let me know what you think :smile: 

Oh I also updated the stackage to use the latest LTS, I can revert if you want.

nalchevanidze commented:
@AlistairB sorry for long delay . there is some merge conflicts when you resolve them feel free to merge it (is approved!).

yeah. would be great if you could also update change-log :)

AlistairB commented on morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs:
```diff
@@ -0,0 +1,88 @@
+    "test custom Errors"
+    ( Left
+        ( FetchErrorProducedErrors
+            ((GQLError "Failure" [Position {line = 3, column = 7}] Nothing Nothing) :| [])
```
The biggest point of uncertainty was here where in master the `GQLError` data constructor is no longer being exported.

I have updated `Data.Morpheus.Types.Internal.AST.Error` to export it again. The other alternative would be to export `Msg (msg)` and `at` then you could use those to build the same `GQLError`.

Can you advise which of these you prefer (or perhaps some other solution)?

nalchevanidze replied:
Hi. please do not expose GQLError.  

perferded approach is to use

```hs
myError :: GQLError
myError = "my error"
```
or 
```hs
myError :: GQLError
myError = "my error" `at` (Position 1 2) 
 ```

nalchevanidze replied:
of course then we should use Pragma "OverloadedStrings" for that.

AlistairB reviewed (commented):
Thanks for the approval @nalchevanidze . I have rebased and updated this PR.

There was one area I was uncertain about which I have left a comment. Also the tests for `morpheus-graphql-app` are failing in master, but otherwise I think everything is passing. (And the Deploy Docs build is failing, apparently unrelated to the changes).

nalchevanidze commented on morpheus-graphql-client/test/Case/JSON/Custom/Errors.hs:
```diff
@@ -0,0 +1,88 @@
+    "test custom Errors"
+    ( Left
+        ( FetchErrorProducedErrors
+            (GQLError "Failure" [Position {line = 3, column = 7}] Nothing Nothing :| [])
```
pleae ue : 
```hs
FetchErrorProducedErrors (("Failure"  `at` Position {line = 3, column = 7}) :| [ ] )
```

nalchevanidze replied:
i thinks , this way is also easier to read :)

nalchevanidze commented on morpheus-graphql-core/src/Data/Morpheus/Types/Error.hs:
```diff
@@ -0,0 +1,16 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Data.Morpheus.Types.Error
```
do we really need this ? .  i think having AST is sufficent

AlistairB replied:
Hi, I have removed this module. I think the downside is that this means users of the client library will need to import `Data.Morpheus.Types.Internal.AST` if they want to inspect the `GQLError` type (unless I am missing something). I think inspecting the error returned is pretty normal as you may want to branch based on what error you got.

I think ideally you should not need to import an `Internal` module for what I think is a common use case. Which isn't to say you need `Data.Morpheus.Types.Error`, you could also choose to export `GQLError (message, position)` from other existing modules.

Anyway, I'm not too fussed either way :)

nalchevanidze replied:
maybe later :) for now is ok how  it is.

nalchevanidze commented on morpheus-graphql-core/src/Data/Morpheus/Types/Internal/AST/Error.hs:
```diff
@@ -15,7 +15,8 @@ module Data.Morpheus.Types.Internal.AST.Error
     GQLErrors,
     GQLError
-      ( message,
+      ( GQLError,
```
please hide GQLError constructor again, users should use new approach of the error generating.

on the second hand i have flexibility to extend error type how i wish.
